### PR TITLE
add menu to add/delete members from memberlist

### PR DESCRIPTION
**Pull request by mightymop**
[xep-0045 9.5 Modifying the member list](https://xmpp.org/extensions/xep-0045.html#modifymember)

### CODE REVIEW

**Comment by sualko** [comment]
Thanks for all your work. It's awesome. I'm currently busy, but I will review them on / short after Christmas.

**Comment by mightymop** [comment]
> 
> 
> Thanks for all your work. It's awesome. I'm currently busy, but I will review them on / short after Christmas.

super...

btw. Wir können auch auf Deutsch quatschen... Muttersprache is mir lieber als Englisch :D

File: src/ui/dialogs/multiUserMemberlist.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+   let content = multiUserMemberlist({});
+
+   let dialog = new Dialog(content);
+   let dom = dialog.open();
+   let connection = multiUserContact.getAccount().getConnection();
+   let items;
+
+   dom.find('form').on('submit', (ev) => {
+      ev.preventDefault();
+      let saveitems = $('.jsxc-memberlist-textarea').val().toString().split('\n');
```
Comment (sualko) [inline src/ui/dialogs/multiUserMemberlist.ts:18]: Trim all at once: `.map(item => item.trim())`

File: src/ui/dialogs/multiUserMemberlist.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+          let found=false;
+          for (;n<saveitems.length;n++)
+          {
+              if (saveitems[n].trim().length>0&&$(items[i]).attr('jid')===saveitems[n].trim())
+              {
+                  found=true;
+                  break;
+              }
+          }
+          if (!found)
```
Comment (sualko) [inline src/ui/dialogs/multiUserMemberlist.ts:33]: No need to loop through all items. Just use `saveItems.indexOf(items[i]) < 0`.

**Review by sualko — CHANGES_REQUESTED** [review]

File: src/ui/dialogs/multiUserMemberlist.ts (hunk 2 above)
Comment (mightymop) [reply src/ui/dialogs/multiUserMemberlist.ts:33]: saveitems is a string array and items is a list of xml objects... dont think indexOf will do the job!?

File: src/ui/dialogs/multiUserMemberlist.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+   let dialog = new Dialog(content);
+   let dom = dialog.open();
+   let connection = multiUserContact.getAccount().getConnection();
+   let items;
+
+   dom.find('form').on('submit', (ev) => {
+      ev.preventDefault();
+      let saveitems = $('.jsxc-memberlist-textarea').val().toString().split('\n');
+      let i=0;
```
[the remark this answers is not in the document]
Comment (mightymop) [reply src/ui/dialogs/multiUserMemberlist.ts]: now in header

File: src/ui/dialogs/multiUserMemberlist.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+              }
+          }
+          if (!found)
+          {
+              items.push($('<item affiliation="member" jid="'+saveitems[i].trim()+'" />')[0]);
+          }
+      }
+
+      connection.getMUCService().setMemberlistMultiUserRoom(multiUserContact.getJid(),items).then(function(stanza){
+           console.log(stanza);
```
[the remark this answers is not in the document]
Comment (mightymop) [reply src/ui/dialogs/multiUserMemberlist.ts]: removed

File: src/ui/dialogs/multiUserMemberlist.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    connection.getMUCService().getMemberlistMultiUserRoom(multiUserContact.getJid())
+      .then(function(stanza){
+
+          if ($(stanza).attr('type')==='result')
+          {
+              items = $(stanza).find('item');
+
+              let textfield = $('#jsxc-memberlist-textarea');
+              textfield.empty();
```
[the remark this answers is not in the document]
Comment (mightymop) [reply src/ui/dialogs/multiUserMemberlist.ts]: took your suggestion

File: src/ui/dialogs/multiUserMemberlist.ts (hunk 1 above)
Comment (mightymop) [reply src/ui/dialogs/multiUserMemberlist.ts]: ok

File: src/ui/dialogs/multiUserMemberlist.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+              for (;i<items.length;i++)
+              {
+                  textfield.append($(items[i]).attr('jid')+'\n');
+              }
+          }
+          else
+          {
+              alert(Translation.t('UNKNOWN_ERROR'));
+          }
+    });
```
[the remark this answers is not in the document]
Comment (mightymop) [reply src/ui/dialogs/multiUserMemberlist.ts]: now dialog only opens on success

File: src/ui/dialogs/multiUserMemberlist.ts (hunk 2 above)
Comment (sualko) [reply src/ui/dialogs/multiUserMemberlist.ts:54]: `items[i].attr('id') && saveItems.indexOf(items[i].attr('id')) < 0` should be the same as the inner loop.

**Comment by sualko** [comment]
Since we only have to send the delta, I was able to simplify the code a bit. Please test and report any issues.

**Review by sualko — APPROVED** [review]

**Comment by mightymop** [comment]
> 
> 
> Since we only have to send the delta, I was able to simplify the code a bit. Please test and report any issues.

i have nothing to add... looks good to me now

**Pull request merged**